XX_BubbleSort/main.cpp

```diff
@@ -0,0 +1,449 @@
+				const string WorkgroupSizeAsStr = std::to_string(WorkgroupSize);
+				const IShaderCompiler::SPreprocessorOptions::SMacroDefinition WorkgroupSizeDefine = { "WORKGROUP_SIZE",WorkgroupSizeAsStr };
+
+				CHLSLCompiler::SOptions options = {};
+				// really we should set it to `ESS_COMPUTE` since we know, but we'll test the `#pragma` handling fur teh lulz
+				options.stage = asset::IShader::E_SHADER_STAGE::ESS_UNKNOWN;
+				// want as much debug as possible
+				options.debugInfoFlags |= IShaderCompiler::E_DEBUG_INFO_FLAGS::EDIF_LINE_BIT;
+				// this lets you source-level debug/step shaders in renderdoc
+				if (physDev->getLimits().shaderNonSemanticInfo)
+					options.debugInfoFlags |= IShaderCompiler::E_DEBUG_INFO_FLAGS::EDIF_NON_SEMANTIC_BIT;
+				// if you don't set the logger and source identifier you'll have no meaningful errors
+				options.preprocessorOptions.sourceIdentifier = "embedded.comp.hlsl";
+				options.preprocessorOptions.logger = m_logger.get();
+				options.preprocessorOptions.extraDefines = { &WorkgroupSizeDefine,&WorkgroupSizeDefine + 1 };
+				if (!(evenPassCPUShader = compiler->compileToSPIRV(evenPass, options)))
+					return logFail("Failed to compile following HLSL Shader:\n%s\n", evenPass);
+				if (!(oddPassCPUShader = compiler->compileToSPIRV(oddPass, options)))
+					return logFail("Failed to compile following HLSL Shader:\n%s\n", oddPass);
+			}
+
+			// Note how each ILogicalDevice method takes a smart-pointer r-value, so that the GPU objects refcount their dependencies
+			smart_refctd_ptr<nbl::video::IGPUShader> evenPassShader = device->createShader(evenPassCPUShader.get());
+			if (!evenPassShader)
+				return logFail("Failed to create a GPU Shader, seems the Driver doesn't like the SPIR-V we're feeding it!\n");
+
+			smart_refctd_ptr<nbl::video::IGPUShader> oddPassShader = device->createShader(oddPassCPUShader.get());
+			if (!oddPassShader)
+				return logFail("Failed to create a GPU Shader, seems the Driver doesn't like the SPIR-V we're feeding it!\n");
+
+			// the simplest example would have used push constants and BDA, but RenderDoc's debugging of that sucks, so I'll demonstrate "classical" binding of buffers with descriptors
+			nbl::video::IGPUDescriptorSetLayout::SBinding bindings[1] = {
+				{
+					.binding = 0,
+					.type = nbl::asset::IDescriptor::E_TYPE::ET_STORAGE_BUFFER,
+					.createFlags = IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE, // not is not the time for descriptor indexing
+					.stageFlags = IGPUShader::ESS_COMPUTE,
+					.count = 1,
+					.samplers = nullptr // irrelevant for a buffer
+				}
+			};
+			smart_refctd_ptr<IGPUDescriptorSetLayout> dsLayout = device->createDescriptorSetLayout(bindings);
+			if (!dsLayout)
+				return logFail("Failed to create a Descriptor Layout!\n");
+
+			const nbl::asset::SPushConstantRange pcRange = { .stageFlags = IShader::ESS_COMPUTE,.offset = 0,.size = sizeof(PushConstantData) };
```

Comment:
Renderdoc now supports BDA, so please use BDA and stick the buffer address in the Push Constant, this way you don't need to create descriptor sets, etc.